client/client.go

```diff
@@ -207,7 +212,11 @@ func (c *Client) Close() error {
 // and receives a unique ID from the agent. The given ID is used to distinguish
 // different clients.
 func (c *Client) Activate(ctx context.Context) error {
-	if c.status == activated {
+	// Apply critical section for whole Activate
+	defer c.clientMutex.Unlock()
+	c.clientMutex.Lock()
+
+	if c.IsActive() {
 		return nil
 	}
```

Comment:
```suggestion
	if c.IsActive() {
		return nil
	}
	
        defer c.clientMutex.Unlock()
	c.clientMutex.Lock()
```
In my opinion, it is okay to set a read-lock in the c.IsActive() method, and then put a write-lock on the logic that changes the resource afterward. 
How about setting up your scope like this?